packages/grid/_modules_/grid/components/DensityPicker.tsx

```diff
@@ -0,0 +1,152 @@
+import { useGridState } from '../hooks/features/core/useGridState';
+
+interface DensityConfig {
+  density: Density;
+  headerHeight: number;
+  rowHeight: number;
+}
+
+interface DensityOption {
+  icon: React.ReactElement;
+  label: DensityTypes;
+}
+
+const DENSITY_DIVISOR = 2;
+const DENSITY_MULTIPLIER = 1.5;
+
+export const DensityPicker = React.memo(function DensityPicker() {
+  const api = React.useContext(ApiContext);
+  const options = useGridSelector(api, optionsSelector);
+  const [, setGridState, forceUpdate] = useGridState(api!);
+  const initialDensityConfig = React.useRef<DensityConfig>({
+    density: options.density,
+    headerHeight: options.headerHeight,
+    rowHeight: options.rowHeight,
+  });
+  const [anchorEl, setAnchorEl] = React.useState(null);
+  const icons = useIcons();
+
+  const DensityShortIcon = icons!.densityShort!;
+  const DensityMediumIcon = icons!.densityMedium!;
+  const DensityTallIcon = icons!.densityTall!;
+
+  const DensityOptions: Array<DensityOption> = [
+    {
+      icon: <DensityShortIcon />,
+      label: DensityTypes.Short,
+    },
+    {
+      icon: <DensityMediumIcon />,
+      label: DensityTypes.Medium,
+    },
+    {
+      icon: <DensityTallIcon />,
+      label: DensityTypes.Tall,
+    },
+  ];
+
+  const getUpdatedDensityConfig = (newDensity: Density): DensityConfig => {
+    switch (newDensity) {
+      case DensityTypes.Short:
+        return {
+          density: newDensity,
+          headerHeight: Math.floor(initialDensityConfig.current.headerHeight / DENSITY_DIVISOR),
+          rowHeight: Math.floor(initialDensityConfig.current.rowHeight / DENSITY_DIVISOR),
+        };
+      case DensityTypes.Tall:
+        return {
+          density: newDensity,
+          headerHeight: Math.floor(initialDensityConfig.current.headerHeight * DENSITY_MULTIPLIER),
+          rowHeight: Math.floor(initialDensityConfig.current.rowHeight * DENSITY_MULTIPLIER),
+        };
+      default:
+        return {
+          ...initialDensityConfig.current,
+          density: newDensity,
+        };
+    }
+  };
+
+  const handleDensityPickerOpen = (event) => setAnchorEl(event.currentTarget);
+  const handleDensityPickerClose = () => setAnchorEl(null);
+
+  const handleUpdateDensity = (newDensity: DensityTypes) => {
+    setAnchorEl(null);
+    setGridState((oldState) => ({
+      ...oldState,
+      options: {
+        ...oldState.options,
+        ...getUpdatedDensityConfig(newDensity),
+      },
+    }));
+    forceUpdate();
+  };
+
+  React.useEffect(() => {
+    if (initialDensityConfig.current.density !== DensityTypes.Medium) {
+      setGridState((oldState) => ({
+        ...oldState,
+        options: {
+          ...oldState.options,
+          ...getUpdatedDensityConfig(initialDensityConfig.current.density),
+        },
+      }));
+      forceUpdate();
+    }
+  }, [setGridState, forceUpdate]);
+
+  const renderDensityOptions: Array<React.ReactElement> = DensityOptions.map((option, index) => (
```

Comment:
We have 2 distincts components here.
The button, in the toolbar 
The menu that let you set the density
Then we should have a useDensityPicker to glue them using the apiRef.
So => button => apiRef.current.showDensityPicker => MenuDisplayed => 
apiRef.current.setDensity('normal') => setRowHeight(options.rowHeight * 1) 🤔 
apiRef.current.hideDensityPicker

This way the button is independant to the menu, and user can render the menu without the button or render their own component and use the api to change the density...
Then we should have an option to enableDensityPicker?

The flow for feature is Component calls apiRef that is on a dedicated hook, the hook update the state and trigger a rerender

Reply:
I agree with how the functionality should he chained but IMO the button and the menu should go together. If you hide the button the menu should also be removed from the Grid and them the user has to implement both, I think that this is expected dev flow. What i can do is move the `Menu` config into its own component but that seems like an unnecessary abstraction since we are already using it from MUI.

Reply:
I would propose we start with the simplest possible option. It's still unclear if developers will use it. Or what they will need in terms of customization power. I think that deferring would be better.